### PR TITLE
Feat/metrics

### PR DESCRIPTION
**Pull request by iGxnon**
Reopen of #521 

Base on #642 

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

**Comment by liangyuanpeng** [comment]
@iGxnon 



I'm trying call the `/metrics` on the port of 2379, and just got the error : `curl: (1) Received HTTP/0.9 when not allowed`

I noticed this PR, so the reason is because xline doesn't implement metrics API yet,

Is this possible release in 0.7.0? Just need the pull mode.

**Comment by iGxnon** [comment]
> Is this possible release in 0.7.0?

Yes, we plan to release it in 0.7.0

> Just need the pull mode.

Push mode is also implemented in this PR.

File: curp/src/server/metrics.rs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+fn meter() -> &'static Meter {
+    METRICS_METER.get_or_init(|| {
+        meter_with_version(
+            env!("CARGO_PKG_NAME"),
+            Some(env!("CARGO_PKG_VERSION")),
+            Some(env!("CARGO_PKG_REPOSITORY")),
+            Some(vec![KeyValue::new("component", "curp")]),
+        )
+    })
+}
```
Comment (Phoenix500526) [inline curp/src/server/metrics.rs:34]: These code are very similar with those code define in other metric.rs files. Shall we define a macro to reduce the duplication?

File: curp/src/server/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    // /// The total number of bytes send to peers.
+    // pub(super) peer_sent_bytes_total: Counter<u64>,
+    // /// The total number of bytes received from peers.
+    // pub(super) peer_received_bytes_total: Counter<u64>,
+    // /// The total number of send failures to peers.
+    // pub(super) peer_sent_failures_total: Counter<u64>
+    // /// The total number of receive failures from peers.
+    // pub(super) peer_received_failures_total: Counter<u64>
+    // /// The round-trip-time histogram between peers.
+    // pub(super) peer_round_trip_time_seconds: Histogram<u64>
```
Comment (Phoenix500526) [inline curp/src/server/metrics.rs:69]: Why do we comment these code?

File: curp/src/server/raw_curp/mod.rs [hunk 3]
```diff
@@ -1342,6 +1395,10 @@ impl<C: Command, RC: RoleChange> RawCurp<C, RC> {
             self.ctx.role_change.on_calibrate();
         }
         st.term = term;
+        if st.role == Role::Leader {
+            // a leader fallback into the follower
+            metrics::get().leader_changes.add(1, &[]);
+        }
```
Comment (Phoenix500526) [inline curp/src/server/raw_curp/mod.rs:1401]: We can merge this block of code with lines 1393 ~ 1396

File: engine/src/proxy.rs [hunk 4]
```diff
@@ -125,7 +127,7 @@ pub enum Snapshot {
     /// Memory snapshot
     Memory(MemorySnapshot),
```
Comment (Phoenix500526) [inline engine/src/proxy.rs:128]: Shall we wrap `MemorySnapshot` into `metrics::Layer`?

Comment (iGxnon) [reply engine/src/proxy.rs:144]: `MemorySnapshot` is currently only used in tests and is not necessary.

File: xline/src/main.rs [hunk 5]
```diff
@@ -143,34 +143,40 @@ use clap::Parser;
 use curp::members::{get_cluster_info_from_remote, ClusterInfo};
 use itertools::Itertools;
 use jsonwebtoken::{DecodingKey, EncodingKey};
-use opentelemetry::{global, runtime::Tokio, sdk::propagation::TraceContextPropagator};
+use opentelemetry::global;
 use opentelemetry_contrib::trace::exporter::jaeger_json::JaegerJsonExporter;
+use opentelemetry_otlp::WithExportConfig;
+use opentelemetry_sdk::runtime::Tokio;
+use opentelemetry_sdk::{metrics::MeterProvider, propagation::TraceContextPropagator};
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply xline/src/main.rs]: It seems that `opentelemetry` only has the basic abstraction and components, `Tokio` and `TraceContextPropagator` are only exported in `opentelemetry_sdk`

File: xline/src/main.rs [hunk 6]
```diff
@@ -143,34 +143,40 @@ use clap::Parser;
...
 use tokio::fs;
 use tracing::{debug, error, info};
 use tracing_appender::non_blocking::WorkerGuard;
 use tracing_subscriber::{fmt::format, prelude::*};
+use utils::config::{default_compact_timeout, MetricsConfig, MetricsPushProtocol};
 use utils::{
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply xline/src/main.rs]: done

File: xline/src/main.rs [hunk 7]
```diff
@@ -576,6 +615,7 @@ async fn main() -> Result<()> {
     );
     debug!("{:?}", server);
     let handle = server.start(server_addr, db_proxy, key_pair).await?;
+    init_metrics(metrics_config)?;
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply xline/src/main.rs]: Moved

File: utils/src/config.rs [hunk 8]
```diff
@@ -828,6 +832,155 @@ impl AuthConfig {
     }
 }
 
+/// Xline log rotation strategy
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply utils/src/config.rs]: Done

File: xline/src/main.rs [hunk 9]
```diff
@@ -194,8 +200,30 @@ struct ServerArgs {
...
+    /// Whether to enable metrics
+    #[clap(long, default_value_t = default_metrics_enable())]
+    metrics_enable: bool,
+    /// Metrics port, default to "9100"
+    #[clap(long, default_value_t = default_metrics_port())]
+    metrics_port: u16,
+    /// Metrics path, default to "/metrics"
+    #[clap(long, default_value_t = default_metrics_path())]
+    metrics_path: String,
+    /// Whether to enable metrics
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply xline/src/main.rs]: Done

File: xline/src/main.rs [hunk 10]
```diff
@@ -597,5 +637,73 @@ async fn main() -> Result<()> {
...
+    Ok(())
+}
+
+/// Start metrics server
+fn init_metrics(config: &MetricsConfig) -> Result<()> {
+    if !config.enable() {
+        return Ok(());
+    }
+    if *config.push() {
+        debug!(
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply xline/src/main.rs]: Done

File: curp/src/server/raw_curp/mod.rs (hunk 3 above)
Comment (iGxnon) [reply curp/src/server/raw_curp/mod.rs]: Done

File: curp/src/server/metrics.rs (hunk 1 above)
Comment (iGxnon) [reply curp/src/server/metrics.rs:34]: It seems a bit tricky to handle. This piece of code is shared across multiple crates, requiring a separate crate to provide the macros.

File: engine/src/metrics.rs [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+
+    /// Clean files of current snapshot
+    async fn clean(&mut self) -> io::Result<()> {
+        self.engine.clean().await
+    }
+}
+
+/// The backend metrics
+static METRICS: OnceLock<Metrics> = OnceLock::new();
+
```
Comment (Phoenix500526) [inline engine/src/metrics.rs:160]: Why don't we need a `METRICS_METER` here?

Comment (iGxnon) [reply engine/src/metrics.rs]: fixed

**Review by themanforfree — APPROVED** [review]

**Review by Phoenix500526 — APPROVED** [review]
LGTM

**Pull request merged**